src/app/pages/admin-panel/admin-panel.component.ts

```diff
@@ -0,0 +1,167 @@
+
+  constructor(
+    private firestore: Firestore,
+    private tba: TheBlueAllianceService
+  ) {
+    this.tba.getEvents().then((events) => {
+      this.events = events.filter(
+        (event) => new Date(event.end_date).getFullYear() > 2019
+      )
+    })
+  }
+
+  async getCurrentData() {
+    let schema = await getDoc(doc(this.firestore, "admin/schema"))
+    let data = schema.data()
+    for (let section of data?.sections) {
+      this.sections.push(section)
+    }
+  }
+
+  async getTeams(game: number): Promise<string[]> {
+    await this.tba.getTeams(this.event, this.stage, game).then((val) => {
+      this.teams = val
+    })
+    return this.teams
+  }
+
+  async getScoutsNames(): Promise<string[]> {
+    const document = await getDocs(collection(this.firestore, this.event))
+    const data = document.docs
+    let names: any[] = []
+    for (let name of data) {
+      if (name.id.includes(this.stage)) names.push(name.id)
+    }
+    return names
+  }
+
+  async findMissingScouts() {
+    this.missingTeams = []
+    this.incorrectTeams = []
+    const scouts = await this.getScoutsNames()
+    let splitedScouts = []
+    for (let scout of scouts) {
+      splitedScouts.push(scout.split(" "))
+    }
+    let gameNumbers = []
+    for (let number of splitedScouts) {
+      gameNumbers.push(Number(number[1]))
+    }
+    for (let i = 1; i < this.findMaxGameNum(gameNumbers) + 1; i++) {
+      let specificTeams = []
+      for (let scout of splitedScouts) {
+        if (Number(scout[1]) === i) {
+          specificTeams.push(scout[2])
+        }
+      }
+
+      const needTeams: string[] = await this.getTeams(i)
+
+      for (let team of needTeams) {
+        if (!specificTeams.includes(team)) {
+          this.missingTeams.push(this.stage + " " + i + " " + team)
+        }
+      }
+
+      for (let team of specificTeams) {
+        if (!needTeams.includes(team)) {
+          this.incorrectTeams.push(this.stage + " " + i + " " + team)
+        }
+      }
+    }
+    this.showMissingTeams = true
+  }
+
+  findMaxGameNum(gameNumbers: number[]): number {
+    return Math.max(...gameNumbers)
+  }
+
+  addSection() {
+    this.sections.push({ title: "", prefix: "", widgets: [] })
+  }
+
+  addWidget(section: section) {
+    section.widgets.push({ key: "", label: "", type: "" })
+  }
+
+  removeWidget(widgetIndex: number, sectionIndex: number) {
+    this.sections[sectionIndex].widgets.splice(widgetIndex, 1)
+  }
+
+  removeSection(sectionIndex: number) {
+    this.sections.splice(sectionIndex, 1)
+  }
+
+  eventToString(event: Event): string {
+    return (event.target as HTMLInputElement).value
+  }
+
+  eventToNumber(event: Event): number {
+    return Number((event.target as HTMLInputElement).value)
+  }
+
+  async update() {
+    this.schema = { sections: this.sections }
+    let stringSchema = JSON.parse(JSON.stringify(this.schema))
+    await setDoc(
+      doc(this.firestore, "admin/schema1"),
+      Object.assign({}, stringSchema)
```

Comment:
What is all of this JSON parse stringily Object assign misery? Just set the doc to this.schema